plugin/product.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package plugin
+
+import (
+	"errors"
+	"io"
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/v6/model"
+)
+
+// ProductHooks is a subset of Hooks
+type ProductHooks interface {
+	OnConfigurationChange() error
+	MessageWillBePosted(ctx *Context, post *model.Post) (*model.Post, string)
+	MessageWillBeUpdated(ctx *Context, newPost, oldPost *model.Post) (*model.Post, string)
+	OnPluginClusterEvent(ctx *Context, ev model.PluginClusterEvent)
+	OnWebSocketDisconnect(webConnID, userID string)
+	OnWebSocketConnect(webConnID, userID string)
+	WebSocketMessageHasBeenPosted(webConnID, userID string, req *model.WebSocketRequest)
+}
+
+type registeredProduct struct {
+	productID   string
+	implemented map[int]struct{}
+	adapter     Hooks
+}
+
+func (rp *registeredProduct) Implements(hookId int) bool {
+	_, ok := rp.implemented[hookId]
+	return ok
+}
+
+type hooksAdapter struct {
+	productHooks ProductHooks
+}
+
+func newRegisteredProduct(pluginID string, productHooks ProductHooks) *registeredProduct {
+	return &registeredProduct{
+		productID: pluginID,
+		implemented: map[int]struct{}{
+			OnConfigurationChangeID:         {},
+			MessageWillBePostedID:           {},
+			MessageWillBeUpdatedID:          {},
+			OnPluginClusterEventID:          {},
+			OnWebSocketConnectID:            {},
+			OnWebSocketDisconnectID:         {},
+			WebSocketMessageHasBeenPostedID: {},
+		},
+		adapter: &hooksAdapter{
+			productHooks: productHooks,
+		},
+	}
+}
+
+func (a *hooksAdapter) OnActivate() error {
+	return errors.New("not implemented")
+}
+
+func (a *hooksAdapter) Implemented() ([]string, error) {
+	return nil, errors.New("not implemented")
+}
+
+func (a *hooksAdapter) OnDeactivate() error {
+	return errors.New("not implemented")
+}
```

Comment:
Does this imply a plugin must implement these hooks?

Reply:
Did you mean a product? If so, not necessarily. For now a product should only implement `ProductHooks` interface. Having said that, [MM-44955](https://mattermost.atlassian.net/browse/MM-44955) is going to improve how hooks are going to be registered. So a product will be free to implement whatever hook they want.